c_src/enif_jq.c

```diff
@@ -65,59 +83,82 @@ static ERL_NIF_TERM make_ok_return(ErlNifEnv* env, ERL_NIF_TERM result) {
 
 static ERL_NIF_TERM parse_nif(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
 {
-    char jv_filter[MAX_JQ_FILTER_LEN] = {0};
+    ERL_NIF_TERM error_msg_bin; 
     // ----------------------------- init --------------------------------------
     jq_state *jq = NULL;
     int ret = JQ_ERROR_UNKNOWN;
     ERL_NIF_TERM ret_term;
-    memset(err_msg, 0, MAX_ERR_MSG_LEN * sizeof(char));
     int dumpopts = 512; // JV_PRINT_SPACE1
     jq = jq_init();
     if (jq == NULL) {
         ret = JQ_ERROR_SYSTEM;
-        strcpy(err_msg, "jq_init: malloc error");
+        const char* error_message = "jq_init: Could not initialize jq";
+        error_msg_bin =
+            make_error_msg_bin(env, error_message, strlen(error_message));
         goto out;
     }
-    jq_set_error_cb(jq, err_callback, err_msg);
+    NifEnvAndErrBinPtr env_and_msg_bin = {
+        .env = env,
+        .error_msg_bin_ptr = &error_msg_bin
+    };
+    jq_set_error_cb(jq, err_callback, &env_and_msg_bin);
 
     // --------------------------- read args -----------------------------------
-    ErlNifBinary erl_jq_filter, erl_json_text;
+    ErlNifBinary erl_jq_filter;
+    ErlNifBinary erl_json_text;
     if (!enif_inspect_binary(env, argv[0], &erl_jq_filter) ||
         !enif_inspect_binary(env, argv[1], &erl_json_text)) {
         ret = JQ_ERROR_BADARG;
-        strcpy(err_msg, "invalid input or filter");
+        const char* error_message = "Expected arguments of type binary but got something else";
```

Comment:
Thanks for the feedback. I don't think it is worth it as this will most likely show up as a parsing or compilation error anyway but I can add test-cases for these cases.

Reply:
Added test case: [cf8e279](https://github.com/emqx/jq/pull/13/commits/cf8e27941d2de1e1d3aee3c9ecbd2c38593fa342)